tests/test_docker_utils.py

```diff
@@ -136,46 +140,56 @@ def tearDown(self):
         self.assertEqual(self.initial_containers, self.client_wrapper.list())
         self.assertEqual(self.initial_tmp, self.ls_tmp())
 
-    # Tests at the top are low level;
-    # Tests at the bottom are at higher levels of abstraction.
-
     def test_at_a_minimum(self):
         # A no-op, but if the tests stall, it may be
         # helpful to see if they are even starting.
         self.assertTrue(True)
 
     def test_container_spec_no_input(self):
-        url = self.client_wrapper.run(DockerContainerSpec(
-            image_name='nginx:1.10.3-alpine',
-            container_name=self.timestamp(),
-            labels={self.test_label: 'true'}
-        ))
+        url = self.get_docker_url()
         self.assert_loads_eventually(url, 'Welcome to nginx!')
 
     def test_container_spec_with_input(self):
-        url = self.client_wrapper.run(DockerContainerSpec(
-            image_name='nginx:1.10.3-alpine',
-            container_name=self.timestamp(),
-            labels={self.test_label: 'true'},
-            input={'foo': 'bar'},
-            container_input_path='/usr/share/nginx/html/index.html'
-        ))
+        url = self.get_docker_url({
+            'input': {'foo': 'bar'},
+            'container_input_path': '/usr/share/nginx/html/index.html'
+        })
         self.assert_loads_eventually(url, '{"foo": "bar"}')
 
+    def test_container_spec_cpu_quota(self):
+        # This test is based on timing, so race conditions are likely.
+
+        # This one is not compared against the others,
+        # since behavior varies depending on whether it is in cache or not.
+        with Timer() as warmup:
+            url = self.get_docker_url()
+            self.assert_loads_eventually(url, 'Welcome to nginx!')
+
+        with Timer() as fast_1:
+            url = self.get_docker_url({'cpus': 1})
+            self.assert_loads_eventually(url, 'Welcome to nginx!')
+
+        with Timer() as slow:
+            url = self.get_docker_url({'cpus': 0.01})
+            self.assert_loads_eventually(url, 'Welcome to nginx!')
+
+        with Timer() as fast_2:
+            url = self.get_docker_url({'cpus': 1})
+            self.assert_loads_eventually(url, 'Welcome to nginx!')
+
+        message = 'Less CPU should be slower: ' \
```

Comment:
Code level changes overall look good.

Maybe if these  tests are too inconsistent or if its too much work to diagnose discrepancies between local and Travis envs we could just test that the code assigning the different amounts of `nano_cpus` works and rely on docker to work as it states it should?